core/src/main/java/org/apache/iceberg/actions/BaseCommitService.java

```diff
@@ -0,0 +1,224 @@
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import org.apache.iceberg.Table;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+import org.apache.iceberg.relocated.com.google.common.collect.Queues;
+import org.apache.iceberg.relocated.com.google.common.collect.Sets;
+import org.apache.iceberg.relocated.com.google.common.util.concurrent.ThreadFactoryBuilder;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * An async service which allows for committing multiple file groups as their rewrites complete. The
+ * service also allows for partial-progress since commits can fail. Once the service has been closed
+ * no new file groups should not be offered.
+ *
+ * <p>Specific implementations provide implementations for {@link #commitOrClean(Set)} and {@link
+ * #abortFileGroup(Object)}
+ *
+ * @param <T> abstract type of file group
+ */
+abstract class BaseCommitService<T> implements Closeable {
+  private static final Logger LOG = LoggerFactory.getLogger(BaseCommitService.class);
+
+  private final Table table;
+  private final ExecutorService committerService;
+  private final ConcurrentLinkedQueue<T> completedRewrites;
+  private final ConcurrentLinkedQueue<String> inProgressCommits;
+  private final List<T> committedRewrites;
+  private final int rewritesPerCommit;
+  private final AtomicBoolean running = new AtomicBoolean(false);
+
+  /**
+   * Constructs a {@link BaseCommitService}
+   *
+   * @param table table to perform commit on
+   * @param rewritesPerCommit number of file groups to include in a commit
+   */
+  BaseCommitService(Table table, int rewritesPerCommit) {
+    this.table = table;
+    LOG.info(
+        "Creating commit service for table {} with {} groups per commit", table, rewritesPerCommit);
+    this.rewritesPerCommit = rewritesPerCommit;
+
+    committerService =
+        Executors.newSingleThreadExecutor(
+            new ThreadFactoryBuilder().setNameFormat("Committer-Service").build());
+
+    completedRewrites = Queues.newConcurrentLinkedQueue();
+    committedRewrites = Lists.newArrayList();
+    inProgressCommits = Queues.newConcurrentLinkedQueue();
+  }
+
+  /**
+   * Perform a commit operation on the table for the set of file groups, should cleanup failed file
+   * groups.
+   *
+   * @param batch set of file groups
+   */
+  abstract void commitOrClean(Set<T> batch);
+
+  /**
+   * Clean up a specified file set by removing any files created for that operation, should not
+   * throw any exceptions
+   *
+   * @param group group of files which are not yet committed
+   */
+  abstract void abortFileGroup(T group);
+
+  /** Starts a single threaded executor service for handling file group commits. */
+  public void start() {
+    Preconditions.checkState(
+        running.compareAndSet(false, true), "Rewrite Commit service already started");
```

Comment:
Minor: We should either change `Commit` -> `commit` or drop `Rewrite` altogether. Other messages refer to this class as just commit service. I'd go for shorter messages but it is up to you.